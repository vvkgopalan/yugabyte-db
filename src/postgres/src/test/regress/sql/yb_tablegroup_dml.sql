--
-- Tablegroups
--

-- CREATE DATABASE test_tablegroups

CREATE DATABASE test_tablegroups;
\c test_tablegroups

-- CREATE TABLEGROUP

CREATE TABLEGROUP tg_test1;
CREATE TABLEGROUP tg_test2;

-- CREATE TABLE
-- No primary key
CREATE TABLE tab_nonkey (a INT) TABLEGROUP tg_test1;
\d tab_nonkey
-- Hash partitioned will fail
CREATE TABLE tab_key (a INT PRIMARY KEY) TABLEGROUP tg_test1;
\d tab_key
-- Range primary key
CREATE TABLE tab_range (a INT, PRIMARY KEY (a ASC)) TABLEGROUP tg_test1;
CREATE TABLE tab_range_multicol (a INT, b INT, PRIMARY KEY (a ASC, b DESC)) TABLEGROUP tg_test1;
-- do not use tablegroup
CREATE TABLE tab_nonkey_nogrp (a INT);
CREATE TABLE tab_hash_nogrp (a INT PRIMARY KEY);

-- INSERT
INSERT INTO tab_range (a) VALUES (0), (1), (2);
INSERT INTO tab_range (a, b) VALUES (0, '0'); -- fail
INSERT INTO tab_range_multicol (a, b) VALUES (0, '0'), (1, '1');
INSERT INTO tab_nonkey_nogrp (a) VALUES (0), (1), (2), (3);
INSERT INTO tab_hash_nogrp (a) VALUES (0), (1), (2), (3);

-- SELECT
SELECT * FROM tab_range;
SELECT * FROM tab_range WHERE a = 2;
SELECT * FROM tab_range WHERE n = '0'; -- fail
SELECT * FROM tab_range_multicol;
SELECT * FROM tab_nonkey_nogrp ORDER BY a ASC;
SELECT * FROM tab_hash_nogrp;

BEGIN;
INSERT INTO tab_range VALUES (4);
SELECT * FROM tab_range;
ROLLBACK;
BEGIN;
INSERT INTO tab_range VALUES (5);
COMMIT;
SELECT * FROM tab_range;

INSERT INTO tab_range VALUES (6), (6); -- fail

-- CREATE INDEX

-- table with index in the tablegroup
CREATE TABLE tab_range2 (a INT, b INT);
CREATE INDEX idx_range2 ON tab_range2 (a);
\d tab_range2
INSERT INTO tab_range2 (a, b) VALUES (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5);
EXPLAIN (COSTS OFF) SELECT * FROM tab_range2 WHERE a = 1;
SELECT * FROM tab_range2 WHERE a = 1;
UPDATE tab_range2 SET b = b + 1 WHERE a > 3;
SELECT * FROM tab_range2 ORDER BY a;
DELETE FROM tab_range2 WHERE a > 3;
SELECT * FROM tab_range2 ORDER BY a;

-- table with no tablegroup with index
CREATE TABLE tab_range_nogrp (a INT, b INT, PRIMARY KEY (a ASC));
CREATE INDEX idx_range_nogrp ON tab_range_nogrp (a);

-- table with index in a different tablegroup
CREATE TABLE tab_range_nonkey2 (a INT, b INT) TABLEGROUP tg_test2;
CREATE INDEX idx_range_nonkey2 ON tab_range_nonkey2 (a);
INSERT INTO tab_range_nonkey2 (a, b) VALUES (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5);
EXPLAIN (COSTS OFF) SELECT * FROM tab_range_nonkey2 WHERE a = 1;
SELECT * FROM tab_range_nonkey2 WHERE a = 1;
UPDATE tab_range_nonkey2 SET b = b + 1 WHERE a > 3;
SELECT * FROM tab_range_nonkey2 ORDER BY a;
DELETE FROM tab_range_nonkey2 WHERE a > 3;
SELECT * FROM tab_range_nonkey2 ORDER BY a;

\dt
\di

-- TRUNCATE TABLE

-- truncate tablegroup table with default index
TRUNCATE TABLE tab_range;
SELECT * FROM tab_range;
INSERT INTO tab_range VALUES (4);
SELECT * FROM tab_range;
INSERT INTO tab_range VALUES (1);
INSERT INTO tab_range VALUES (2), (5);
SELECT * FROM tab_range;
DELETE FROM tab_range WHERE a = 2;
SELECT * FROM tab_range;

TRUNCATE TABLE tab_range;

-- truncate non-colocated table without index
TRUNCATE TABLE tab_nonkey_nogrp;
SELECT * FROM tab_nonkey_nogrp;

-- truncate colocated table with explicit index
TRUNCATE TABLE tab_range_nonkey2;
SELECT * FROM tab_range_nonkey2;

\dt
\di

-- ALTER TABLE
INSERT INTO tab_range (a) VALUES (0), (1), (2);
INSERT INTO tab_range_nonkey2 (a, b) VALUES (0, 0), (1, 1);

SELECT * FROM tab_range;
SELECT * FROM tab_range_nonkey2 ORDER BY a;

-- Alter tablegrouped tables
ALTER TABLE tab_range ADD COLUMN x INT;
ALTER TABLE tab_range_nonkey2 DROP COLUMN b;

SELECT * FROM tab_range;
SELECT * FROM tab_range_nonkey2 ORDER BY a;

ALTER TABLE tab_range_nonkey2 RENAME TO tab_range_nonkey2_renamed;
SELECT * FROM tab_range_nonkey2_renamed ORDER BY a;
SELECT * FROM tab_range_nonkey2;

-- DROP TABLE

-- drop table that is in a tablegroup with default index
DROP TABLE tab_range;
SELECT * FROM tab_range;

-- drop non-colocated table without index
DROP TABLE tab_nonkey_nogrp;
SELECT * FROM tab_nonkey_nogrp;

--- drop colocated table with explicit index
DROP TABLE tab_range_nonkey2_renamed;
SELECT * FROM tab_range_nonkey2_renamed;

-- DROP INDEX

-- drop index on colocated table
DROP INDEX idx_range2;
EXPLAIN SELECT * FROM tab_range2 WHERE a = 1;

\dt
\di

-- DROP TABLEGROUP
DROP TABLEGROUP tg_test1; -- fail
DROP TABLEGROUP tg_test2;

-- drop database
\c yugabyte
DROP DATABASE test_tablegroups;
\c test_tablegroups
