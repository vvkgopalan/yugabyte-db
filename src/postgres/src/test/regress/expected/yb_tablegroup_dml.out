--
-- Tablegroups
--
-- CREATE DATABASE test_tablegroups
CREATE DATABASE test_tablegroups;
\c test_tablegroups
-- CREATE TABLEGROUP
CREATE TABLEGROUP tg_test1;
CREATE TABLEGROUP tg_test2;
-- CREATE TABLE
-- No primary key
CREATE TABLE tab_nonkey (a INT) TABLEGROUP tg_test1;
\d tab_nonkey
             Table "public.tab_nonkey"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 a      | integer |           |          |

-- Hash partitioned will fail
CREATE TABLE tab_key (a INT PRIMARY KEY) TABLEGROUP tg_test1;
ERROR:  Invalid argument: Invalid table definition: Error creating table test_tablegroups.tab_key on the master: Cannot colocate hash partitioned table
\d tab_key
-- Range primary key
CREATE TABLE tab_range (a INT, PRIMARY KEY (a ASC)) TABLEGROUP tg_test1;
CREATE TABLE tab_range_multicol (a INT, b INT, PRIMARY KEY (a ASC, b DESC)) TABLEGROUP tg_test1;
-- do not use tablegroup
CREATE TABLE tab_nonkey_nogrp (a INT);
CREATE TABLE tab_hash_nogrp (a INT PRIMARY KEY);
-- INSERT
INSERT INTO tab_range (a) VALUES (0), (1), (2);
INSERT INTO tab_range (a, b) VALUES (0, '0'); -- fail
ERROR:  column "b" of relation "tab_range" does not exist
LINE 1: INSERT INTO tab_range (a, b) VALUES (0, '0');
                                  ^
INSERT INTO tab_range_multicol (a, b) VALUES (0, '0'), (1, '1');
INSERT INTO tab_nonkey_nogrp (a) VALUES (0), (1), (2), (3);
INSERT INTO tab_hash_nogrp (a) VALUES (0), (1), (2), (3);
-- SELECT
SELECT * FROM tab_range;
 a
---
 0
 1
 2
(3 rows)

SELECT * FROM tab_range WHERE a = 2;
 a
---
 2
(1 row)

SELECT * FROM tab_range WHERE n = '0'; -- fail
ERROR:  column "n" does not exist
LINE 1: SELECT * FROM tab_range WHERE n = '0';
                                      ^
SELECT * FROM tab_range_multicol;
 a | b
---+---
 0 | 0
 1 | 1
(2 rows)

SELECT * FROM tab_nonkey_nogrp ORDER BY a ASC;
 a
---
 0
 1
 2
 3
(4 rows)

SELECT * FROM tab_hash_nogrp;
 a
---
 1
 0
 2
 3
(4 rows)

BEGIN;
INSERT INTO tab_range VALUES (4);
SELECT * FROM tab_range;
 a
---
 0
 1
 2
 4
(4 rows)

ROLLBACK;
BEGIN;
INSERT INTO tab_range VALUES (5);
COMMIT;
SELECT * FROM tab_range;
 a
---
 0
 1
 2
 5
(4 rows)

INSERT INTO tab_range VALUES (6), (6); -- fail
ERROR:  duplicate key value violates unique constraint "tab_range_pkey"
-- CREATE INDEX
-- table with index in the tablegroup
CREATE TABLE tab_range2 (a INT, b INT);
CREATE INDEX idx_range2 ON tab_range2 (a);
\d tab_range2
             Table "public.tab_range2"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 a      | integer |           |          |
 b      | integer |           |          |
Indexes:
    "idx_range2" lsm (a HASH)

INSERT INTO tab_range2 (a, b) VALUES (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5);
EXPLAIN (COSTS OFF) SELECT * FROM tab_range2 WHERE a = 1;
                QUERY PLAN
-------------------------------------------
 Index Scan using idx_range2 on tab_range2
   Index Cond: (a = 1)
(2 rows)

SELECT * FROM tab_range2 WHERE a = 1;
 a | b
---+---
 1 | 1
(1 row)

UPDATE tab_range2 SET b = b + 1 WHERE a > 3;
SELECT * FROM tab_range2 ORDER BY a;
 a | b
---+---
 0 | 0
 1 | 1
 2 | 2
 3 | 3
 4 | 5
 5 | 6
(6 rows)

DELETE FROM tab_range2 WHERE a > 3;
SELECT * FROM tab_range2 ORDER BY a;
 a | b
---+---
 0 | 0
 1 | 1
 2 | 2
 3 | 3
(4 rows)

-- table with no tablegroup with index
CREATE TABLE tab_range_nogrp (a INT, b INT, PRIMARY KEY (a ASC));
CREATE INDEX idx_range_nogrp ON tab_range_nogrp (a);
-- table with index in a different tablegroup
CREATE TABLE tab_range_nonkey2 (a INT, b INT) TABLEGROUP tg_test2;
CREATE INDEX idx_range_nonkey2 ON tab_range_nonkey2 (a);
INSERT INTO tab_range_nonkey2 (a, b) VALUES (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5);
EXPLAIN (COSTS OFF) SELECT * FROM tab_range_nonkey2 WHERE a = 1;
                       QUERY PLAN
---------------------------------------------------------
 Index Scan using idx_range_nonkey2 on tab_range_nonkey2
   Index Cond: (a = 1)
(2 rows)

SELECT * FROM tab_range_nonkey2 WHERE a = 1;
 a | b
---+---
 1 | 1
(1 row)

UPDATE tab_range_nonkey2 SET b = b + 1 WHERE a > 3;
SELECT * FROM tab_range_nonkey2 ORDER BY a;
 a | b
---+---
 0 | 0
 1 | 1
 2 | 2
 3 | 3
 4 | 5
 5 | 6
(6 rows)

DELETE FROM tab_range_nonkey2 WHERE a > 3;
SELECT * FROM tab_range_nonkey2 ORDER BY a;
 a | b
---+---
 0 | 0
 1 | 1
 2 | 2
 3 | 3
(4 rows)

\dt
               List of relations
 Schema |        Name        | Type  |  Owner
--------+--------------------+-------+----------
 public | tab_hash_nogrp     | table | yugabyte
 public | tab_nonkey         | table | yugabyte
 public | tab_nonkey_nogrp   | table | yugabyte
 public | tab_range          | table | yugabyte
 public | tab_range2         | table | yugabyte
 public | tab_range_multicol | table | yugabyte
 public | tab_range_nogrp    | table | yugabyte
 public | tab_range_nonkey2  | table | yugabyte
(8 rows)

\di
                            List of relations
 Schema |          Name           | Type  |  Owner   |       Table
--------+-------------------------+-------+----------+--------------------
 public | idx_range2              | index | yugabyte | tab_range2
 public | idx_range_nogrp         | index | yugabyte | tab_range_nogrp
 public | idx_range_nonkey2       | index | yugabyte | tab_range_nonkey2
 public | tab_hash_nogrp_pkey     | index | yugabyte | tab_hash_nogrp
 public | tab_range_multicol_pkey | index | yugabyte | tab_range_multicol
 public | tab_range_nogrp_pkey    | index | yugabyte | tab_range_nogrp
 public | tab_range_pkey          | index | yugabyte | tab_range
(7 rows)

-- TRUNCATE TABLE
-- truncate tablegroup table with default index
TRUNCATE TABLE tab_range;
SELECT * FROM tab_range;
 a
---
(0 rows)

INSERT INTO tab_range VALUES (4);
SELECT * FROM tab_range;
 a
---
 4
(1 row)

INSERT INTO tab_range VALUES (1);
INSERT INTO tab_range VALUES (2), (5);
SELECT * FROM tab_range;
 a
---
 1
 2
 4
 5
(4 rows)

DELETE FROM tab_range WHERE a = 2;
SELECT * FROM tab_range;
 a
---
 1
 4
 5
(3 rows)

TRUNCATE TABLE tab_range;
-- truncate non-colocated table without index
TRUNCATE TABLE tab_nonkey_nogrp;
SELECT * FROM tab_nonkey_nogrp;
 a
---
(0 rows)

-- truncate colocated table with explicit index
TRUNCATE TABLE tab_range_nonkey2;
SELECT * FROM tab_range_nonkey2;
 a | b
---+---
(0 rows)

\dt
               List of relations
 Schema |        Name        | Type  |  Owner
--------+--------------------+-------+----------
 public | tab_hash_nogrp     | table | yugabyte
 public | tab_nonkey         | table | yugabyte
 public | tab_nonkey_nogrp   | table | yugabyte
 public | tab_range          | table | yugabyte
 public | tab_range2         | table | yugabyte
 public | tab_range_multicol | table | yugabyte
 public | tab_range_nogrp    | table | yugabyte
 public | tab_range_nonkey2  | table | yugabyte
(8 rows)

\di
                            List of relations
 Schema |          Name           | Type  |  Owner   |       Table
--------+-------------------------+-------+----------+--------------------
 public | idx_range2              | index | yugabyte | tab_range2
 public | idx_range_nogrp         | index | yugabyte | tab_range_nogrp
 public | idx_range_nonkey2       | index | yugabyte | tab_range_nonkey2
 public | tab_hash_nogrp_pkey     | index | yugabyte | tab_hash_nogrp
 public | tab_range_multicol_pkey | index | yugabyte | tab_range_multicol
 public | tab_range_nogrp_pkey    | index | yugabyte | tab_range_nogrp
 public | tab_range_pkey          | index | yugabyte | tab_range
(7 rows)

-- ALTER TABLE
INSERT INTO tab_range (a) VALUES (0), (1), (2);
INSERT INTO tab_range_nonkey2 (a, b) VALUES (0, 0), (1, 1);
SELECT * FROM tab_range;
 a
---
 0
 1
 2
(3 rows)

SELECT * FROM tab_range_nonkey2 ORDER BY a;
 a | b
---+---
 0 | 0
 1 | 1
(2 rows)

-- Alter tablegrouped tables
ALTER TABLE tab_range ADD COLUMN x INT;
ALTER TABLE tab_range_nonkey2 DROP COLUMN b;
SELECT * FROM tab_range;
 a | x
---+---
 0 |
 1 |
 2 |
(3 rows)

SELECT * FROM tab_range_nonkey2 ORDER BY a;
 a
---
 0
 1
(2 rows)

ALTER TABLE tab_range_nonkey2 RENAME TO tab_range_nonkey2_renamed;
SELECT * FROM tab_range_nonkey2_renamed ORDER BY a;
 a
---
 0
 1
(2 rows)

SELECT * FROM tab_range_nonkey2;
ERROR:  relation "tab_range_nonkey2" does not exist
LINE 1: SELECT * FROM tab_range_nonkey2;
                      ^
-- DROP TABLE
-- drop table that is in a tablegroup with default index
DROP TABLE tab_range;
SELECT * FROM tab_range;
ERROR:  relation "tab_range" does not exist
LINE 1: SELECT * FROM tab_range;
                      ^
-- drop non-colocated table without index
DROP TABLE tab_nonkey_nogrp;
SELECT * FROM tab_nonkey_nogrp;
ERROR:  relation "tab_nonkey_nogrp" does not exist
LINE 1: SELECT * FROM tab_nonkey_nogrp;
                      ^
--- drop colocated table with explicit index
DROP TABLE tab_range_nonkey2_renamed;
SELECT * FROM tab_range_nonkey2_renamed;
ERROR:  relation "tab_range_nonkey2_renamed" does not exist
LINE 1: SELECT * FROM tab_range_nonkey2_renamed;
                      ^
-- DROP INDEX
-- drop index on colocated table
DROP INDEX idx_range2;
EXPLAIN SELECT * FROM tab_range2 WHERE a = 1;
                          QUERY PLAN
---------------------------------------------------------------
 Seq Scan on tab_range2  (cost=0.00..102.50 rows=1000 width=8)
   Filter: (a = 1)
(2 rows)

\dt
               List of relations
 Schema |        Name        | Type  |  Owner
--------+--------------------+-------+----------
 public | tab_hash_nogrp     | table | yugabyte
 public | tab_nonkey         | table | yugabyte
 public | tab_range2         | table | yugabyte
 public | tab_range_multicol | table | yugabyte
 public | tab_range_nogrp    | table | yugabyte
(5 rows)

\di
                            List of relations
 Schema |          Name           | Type  |  Owner   |       Table
--------+-------------------------+-------+----------+--------------------
 public | idx_range_nogrp         | index | yugabyte | tab_range_nogrp
 public | tab_hash_nogrp_pkey     | index | yugabyte | tab_hash_nogrp
 public | tab_range_multicol_pkey | index | yugabyte | tab_range_multicol
 public | tab_range_nogrp_pkey    | index | yugabyte | tab_range_nogrp
(4 rows)

-- DROP TABLEGROUP
DROP TABLEGROUP tg_test1; -- fail
ERROR:  tablegroup "tg_test1" is not empty
DROP TABLEGROUP tg_test2;
-- drop database
\c yugabyte
DROP DATABASE test_tablegroups;
\c test_tablegroups
\connect: FATAL:  database "test_tablegroups" does not exist
